src/OrleansTelemetryConsumers.Counters/PerfCounterTelemetryConsumer.cs

```diff
@@ -0,0 +1,224 @@
+            }
+            return ctrCreationData.ToArray();
+        }
+
+        /// <summary>
+        /// Register Orleans perf counters with Windows
+        /// </summary>
+        /// <remarks>Note: Program needs to be running as Administrator to be able to delete Windows perf counters.</remarks>
+        public void InstallCounters()
+        {
+            var collection = new CounterCreationDataCollection();
+            collection.AddRange(GetCounterCreationData());
+
+            PerformanceCounterCategory.Create(
+                CATEGORY_NAME,
+                CATEGORY_DESCRIPTION,
+                PerformanceCounterCategoryType.SingleInstance,
+                collection);
+        }
+
+        /// <summary>
+        /// Delete any existing perf counters registered with Windows
+        /// </summary>
+        /// <remarks>Note: Program needs to be running as Administrator to be able to delete Windows perf counters.</remarks>
+        public void DeleteCounters()
+        {
+            PerformanceCounterCategory.Delete(CATEGORY_NAME);
+        }
+
+        private PerfCounterConfigData GetCounter(string counterName)
+        {
+            return perfCounterData.Where(pcd => GetPerfCounterName(pcd) == counterName).SingleOrDefault();
+        }
+
+        #endregion
+
+        #region IMetricTelemetryConsumer Methods
+
+        public void IncrementMetric(string name) => WriteMetric(name, UpdateMode.Increment);
+
+        public void IncrementMetric(string name, double value) => WriteMetric(name, UpdateMode.Increment, value);
+
+        public void TrackMetric(string name, double value, IDictionary<string, string> properties = null) => WriteMetric(name, UpdateMode.Set, value);
+
+        public void TrackMetric(string name, TimeSpan value, IDictionary<string, string> properties = null) => WriteMetric(name, UpdateMode.Set, value.Ticks);
+
+        public void DecrementMetric(string name) => WriteMetric(name, UpdateMode.Decrement);
+
+        public void DecrementMetric(string name, double value) => WriteMetric(name, UpdateMode.Decrement, value);
+
+        public void Flush() { }
+
+        public void Close() { }
+
+        private void WriteMetric(string name, UpdateMode mode = UpdateMode.Increment, double? value = null)
+        {
+            PerfCounterConfigData cd = GetCounter(name);
+            if (cd == null)
+            {
+                if (logger.IsVerbose) logger.Verbose(ErrorCode.PerfCounterNotFound, "No perf counter found for {0}", name);
+                return;
+            }
+
+            StatisticName statsName = cd.Name;
+            string perfCounterName = GetPerfCounterName(cd);
+
+            try
+            {
+                if (cd.PerfCounter == null)
+                {
+                    if (logger.IsVerbose) logger.Verbose(ErrorCode.PerfCounterUnableToConnect, "No perf counter found for {0}", name);
+                    cd.PerfCounter = CreatePerfCounter(perfCounterName);
```

Comment:
This code must be thread safe now.  Original code was called periodically by a single thread, this code can be called on any Orleans thread.

Suggest allowing no state change (like creating a performance counter) outside the initialization phase.